3_visualize/src/drought_learner_viz_components.R

```diff
@@ -885,5 +891,153 @@ frame_l <- function(p2_streamflow_learner_viz_df, p2_droughts_learner_viz_df,
+    mutate(start_year = year(start),
+           start_noYr = format(as.Date(start), "%m-%d"),
+           start_fakeYr = as.Date(sprintf("1963-%s", start_noYr)),
+           end_year = year(end),
+           end_noYr = format(as.Date(end), "%m-%d"),
+           end_fakeYr = as.Date(sprintf("1963-%s", end_noYr)))
+  p2_droughts_jd <- p2_droughts_jd %>% 
+    mutate(start_year = year(start),
+           start_noYr = format(as.Date(start), "%m-%d"),
+           start_fakeYr = as.Date(sprintf("1963-%s", start_noYr)),
+           end_year = year(end),
+           end_noYr = format(as.Date(end), "%m-%d"),
+           end_fakeYr = as.Date(sprintf("1963-%s", end_noYr)))
+  
+  # Deal with drought events that wrap year
+  #   1. filter out only those 
+  p2_droughts_jd_wrapYr <- p2_droughts_jd %>%
+    filter(start_year != end_year)
+  p2_droughts_site_wrapYr <- p2_droughts_site %>%
+    filter(start_year != end_year)
+  #   2. Duplicate records
+  p2_droughts_jd_wrapYr_beginningOfYear <- p2_droughts_jd_wrapYr %>%
+    mutate(start = as.Date(sprintf("%s-01-01", year(end))))
+  p2_droughts_jd_wrapYr_endOfYear <- p2_droughts_jd_wrapYr %>%
+    mutate(end = as.Date(sprintf("%s-12-31", year(start))))
+  p2_droughts_jd_wrapYr <- bind_rows(p2_droughts_jd_wrapYr_beginningOfYear,
+                                     p2_droughts_jd_wrapYr_endOfYear) 
+  p2_droughts_site_wrapYr_beginningOfYear <- p2_droughts_site_wrapYr %>%
+    mutate(start = as.Date(sprintf("%s-01-01", year(end))))
+  p2_droughts_site_wrapYr_endOfYear <- p2_droughts_site_wrapYr %>%
+    mutate(end = as.Date(sprintf("%s-12-31", year(start))))
+  p2_droughts_site_wrapYr <- bind_rows(p2_droughts_site_wrapYr_beginningOfYear,
+                                       p2_droughts_site_wrapYr_endOfYear) 
+  #   3. Fix metadata
+  p2_droughts_jd_wrapYr <- p2_droughts_jd_wrapYr %>%
+    mutate(start_year = year(start),
+           start_noYr = format(as.Date(start), "%m-%d"),
+           start_fakeYr = as.Date(sprintf("1963-%s", start_noYr)),
+           end_year = year(end),
+           end_noYr = format(as.Date(end), "%m-%d"),
+           end_fakeYr = as.Date(sprintf("1963-%s", end_noYr)))
+  p2_droughts_site_wrapYr <- p2_droughts_site_wrapYr %>%
+    mutate(start_year = year(start),
+           start_noYr = format(as.Date(start), "%m-%d"),
+           start_fakeYr = as.Date(sprintf("1963-%s", start_noYr)),
+           end_year = year(end),
+           end_noYr = format(as.Date(end), "%m-%d"),
+           end_fakeYr = as.Date(sprintf("1963-%s", end_noYr)))
+  #   4. Merge back in with other records
+  p2_droughts_jd_noWrapYr <- p2_droughts_jd %>%
+    filter(start_year == end_year)
+  p2_droughts_jd <- bind_rows(p2_droughts_jd_wrapYr,
+                              p2_droughts_jd_noWrapYr)
+  p2_droughts_site_noWrapYr <- p2_droughts_site %>%
+    filter(start_year == end_year)
+  p2_droughts_site <- bind_rows(p2_droughts_site_wrapYr,
+                                p2_droughts_site_noWrapYr)
+  
```

Comment:
SC: This may well be on your to-do list, but when you refactor, I think all of the code above this line, which is repeated for each dataset, should be pulled into a separate data-prep function that can be called once for each dataset, with the raw dataset, method_name, site, threshold value, and placeholder year (1963) as arguments. Since it is data prep, I think it could potentially also be moved to `2_process.R`. That would a) make the data prep more modular and reduce repeated code, and b) keep this function about generating the plot

Reply:
Added to issue #14